### DISCUSSION TITLE
Thanks and relations between columns with non-enumerable domains.

### DISCUSSION THREAD

**Discussion by DavidPratten** [EvgSkv/logica#24] [Ideas]
Hi Evgeny and Konstantin,

I've been dreaming about the possibility of this language for some time.  Thankyou.

My intuition has been that the declarative relation-based SELECT + VIEW + CTE way of expressing complex algorithms (e.g. graph algorithms) is unique and extremely useful but is held back by the lack of:

- Modularisation and encapsulation of queries
- The ability to capture relations between domains that are not realistically enumerable. 

Logica addresses modularisation and encapsulation of queries.  

An example relation between columns with non-enumerable domains might be 

    FutureValue(InterestRate, NumberOfPeriods, PaymentPerPeriod).  

If such a computed relation was implemented, the developer could SELECT over it, just like any table without being able to tell that this relation was never materialised.

JOINs over such computed relations may require delegation to a constraint solver. So the (under the hood) language for specifying such computed relations should be easily translatable as input into a constraint solver.

Here are my musings on this topic from last month ...  [Table Valued Contrained Functions (TVCF) - Constraint Programming and SQL](https://davidpratten.com/2021/03/22/table-valued-constrained-functions-tvcf-constraint-programming-and-sql/) 

Your thoughts?

David

**Comment by EvgSkv** [comment]
David,
thank you for the supportive feedback and raising this interesting topic!

Logica can work with columns with non-enumerable domains right now, but what it can do with them is limited to what's achievable by _injection_ and _equality resolution_. This is briefly discussed in section [Infinite predicates](https://colab.sandbox.google.com/github/EvgSkv/logica/blob/main/tutorial/Logica_tutorial.ipynb#scrollTo=08a0mLwQBCpb)  of the tutorial. In particular this allows Logica treat regular functions as predicates and helps keep the syntax clean.

In your article you discuss predicate `XPlusZEqualsY`, pointing that it should be able to get Z from X and Y, as well as X from Y and Z.
If you want this functionality in Logica you could actually implement it as follows. I am taking a liberty to use a short name `Add` instead of a longer `XPlusZEqualsY`. I am also using syntactic sugar of Logica implementing it as a function, it would also work as a 3-ary predicate.

```python
# Implementing addition that works in multi-way inference.
Add(x, y) = z :-
  x + y == z,
  z - x == y,
  z - y == x;

# Now we can use Add by setting any pair of variables.
# Compiled SQL would have redundant WHERE clause qualities, but the result
# should be mathematically correct.
P1(a, b, c) :- a in Range(10), b in Range(10), c == Add(a, b);
P2(a, b, c) :- a in Range(10), c in Range(10), c == Add(a, b);
P3(a, b, c) :- b in Range(10), c in Range(10), c == Add(a, b);
```

It is an interesting direction of development of the language to add automatic solvers, which would analyze definitions of predicate, so we could do something like this.

```python
# WARNING: This is a hypothetical feature that enables linear equations solver,
# there is no such annotation at the moment.
@EnableSolver("LinearEquations");

# LinearEquations solver is turned on, so the compiler would figure out that a = b = c = 0.5.
# At the moment this would result in an error stating that no way to assign a, b, c was found.
MyNumbers(a, b, c) :- a + b == 1, a + c == 1, b + c == 1;
```

Let me know if this fits your vision.

**Reply by DavidPratten** [reply]
Hi Evgeny,

Thanks for your quick reply.  I have some ideas and I’ll reply once I’ve had a chance to play with Logica. 

I also look forward to following your next steps!

David
